### PR TITLE
Dispute vote filtering for block authors

### LINKED ISSUE #3472: Disputes Runtime: Filter redundant and ancient votes from inherent MultiDisputeStatementSet
This is the final piece of the puzzle when it comes to wiring the inherent-data pipeline to actually produce `ParasInherent`s that contain dispute votes.

The Provisioner gives _all_ recent disputes to the runtime because of the expectation that disputes are rare. As such, it's very likely that votes are redundant or even correspond to outdated disputes or disputes that concluded long ago.

We need to implement the `filter_multi_dispute_data` function in the disputes module to select only votes which are relevant and don't overfill spam slots. There should be a lot of code we can share between the actual import logic and the filtering logic. Maybe 2 different modes of the `provide_multi_dispute_data` routine.

### PR DESCRIPTION
**Pull request by rphmeier**
Closes #3472 

When the node-side is authoring a block, it passes _all votes_ from _all disputes from recent sessions_ to an instance of the runtime as `InherentData`. The `ProvideInherent` implementation for the `ParasInherent` [here](https://github.com/paritytech/polkadot/blob/master/runtime/parachains/src/paras_inherent.rs#L290) invokes the filtering function, which is intended to detect duplicates and spam and so on.

This pull request actually provides the implementation of that filtering logic, to be used to create `ParasInherent` inherents that actually contain dispute votes so the chain can witness disputes.

### CODE REVIEW

File: runtime/parachains/src/disputes.rs [hunk 1]
```diff
@@ -516,6 +512,47 @@ impl<BlockNumber: Clone> DisputeStateImporter<BlockNumber> {
...
+	{
+		match self {
+			StatementSetFilter::RemoveAll => None,
+			StatementSetFilter::RemoveIndices(mut indices) => {
+				indices.sort();
+				indices.dedup();
+
+				// reverse order ensures correctness
+				for index in indices.into_iter().rev() {
+					statement_set.statements.remove(index);
```
Comment (ordian) [inline runtime/parachains/src/disputes.rs:537]: as usual, if we don't care about the ordering after this operation, I'd suggest to use `swap_remove` which guarantees overall linear complexity

File: runtime/parachains/src/disputes.rs [hunk 2]
```diff
@@ -644,6 +682,181 @@ impl<T: Config> Pallet<T> {
 		Ok(fresh)
 	}
 
+	fn filter_multi_dispute_data(statement_sets: &mut MultiDisputeStatementSet) {
+		let config = <configuration::Pallet<T>>::config();
+
+		let old_statement_sets = sp_std::mem::replace(statement_sets, Vec::new());
```
Comment (ordian) [inline runtime/parachains/src/disputes.rs:688]:
```suggestion
		let old_statement_sets = sp_std::mem::take(statement_sets);
```

File: runtime/parachains/src/disputes.rs [hunk 3]
```diff
@@ -644,6 +682,181 @@ impl<T: Config> Pallet<T> {
...
+			let mut targets = Vec::new();
+			old_statement_sets.into_iter().filter(move |set| {
+				let target = (set.candidate_hash, set.session);
+				let dup = targets.contains(&target);
+				if !dup {
+					targets.push(target);
+				}
+
+				!dup
+			})
```
Comment (ordian) [inline runtime/parachains/src/disputes.rs:701]: another time complexity concern, consider using `BTreeSet` for `targets`

File: runtime/parachains/src/disputes.rs [hunk 4]
```diff
@@ -516,6 +513,48 @@ impl<BlockNumber: Clone> DisputeStateImporter<BlockNumber> {
...
+			StatementSetFilter::RemoveAll => None,
+			StatementSetFilter::RemoveIndices(mut indices) => {
+				indices.sort();
+				indices.dedup();
+
+				// reverse order ensures correctness
+				for index in indices.into_iter().rev() {
+					// swap_remove guarantees linear complexity.
+					statement_set.statements.swap_remove(index);
+				}
```
Comment (drahnr) [inline runtime/parachains/src/disputes.rs:540]: Very elegant :+1:

File: runtime/parachains/src/disputes.rs [hunk 5]
```diff
@@ -644,6 +683,176 @@ impl<T: Config> Pallet<T> {
...
+						filter.remove_index(i);
+						continue
+					}
+					Some(v) => v,
+				};
+
+				let valid = match statement {
+					DisputeStatement::Valid(_) => true,
+					DisputeStatement::Invalid(_) => false,
+				};
```
Comment (drahnr) [inline runtime/parachains/src/disputes.rs:767]: nit: adding a `fn is_...(&self) -> bool` would reduce the clutter a bit.

**Review by drahnr — APPROVED** [review]
very much appreciate the effort put into documenting the rationale of innocent looking code :+1:

**Review by andresilva — APPROVED** [review]

File: runtime/parachains/src/disputes.rs (hunk 4 above)
Comment (rphmeier) [reply runtime/parachains/src/disputes.rs:540]: credit to @ordian

**Comment by rphmeier** [comment]
bot merge

**Comment by ordian** [comment]
(feel free to edit the audit label)

bot merge

**Pull request merged**